Refactored the implementation of RangeSet.

ThouCheese opened the pull request:
Refactored the implementation of RangeSet. The original implementation did not enforce privacy of the fields of `Range`, which meant that the implementation of `Range` could not be easily swapped out. I also added tests for all functions of RangeSet and Range. I also added several extra functions to Range that make the implementation of RangeSet a bit more legible (thing like writing `range1.touches(range2)` instead of `range1.start() <= range2.end() && range1.end() >= range2.start()`). I also favoured operator overloading over method calls, so instead of `range1.minus(range2)` we can write `range1 - range2`. I hope this PR is welcome! Let me know what you think.

Johannesd3 commented:
There are some clippy warnings:

```
warning: This sequence of operators looks suspiciously like a bug.
   --> audio/src/range_set.rs:223:49
    |
223 |             if to_sub.start() <= cur.start() && cur.start() < to_sub.end() {
    |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: I think you meant: `cur.end() < to_sub.end()`
    |
    = note: `#[warn(clippy::suspicious_operation_groupings)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#suspicious_operation_groupings

warning: This sequence of operators looks suspiciously like a bug.
   --> audio/src/range_set.rs:223:49
    |
223 |             if to_sub.start() <= cur.start() && cur.start() < to_sub.end() {
    |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: I think you meant: `cur.end() < to_sub.end()`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#suspicious_operation_groupings

warning: use Vec::sort_by_key here instead
  --> audio/src/range_set.rs:45:9
   |
45 |         ranges.sort_unstable_by(|r1, r2| r1.start().cmp(&r2.start()));
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `ranges.sort_unstable_by_key(|r1| r1.start())`
   |
   = note: `#[warn(clippy::unnecessary_sort_by)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_sort_by

error: suspicious use of binary operator in `Sub` impl
   --> audio/src/range_set.rs:276:14
    |
276 |         self += rhs;
    |              ^^
    |
    = note: `#[deny(clippy::suspicious_arithmetic_impl)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#suspicious_arithmetic_impl

error: aborting due to previous error; 3 warnings emitted

```

ThouCheese commented:
Ah I had seen only the first one, which is a false positive. The last two ones however are not! I fixed this and added `#![allow(clippy::suspicious_operation_groupings)]` to `range_set.rs`.

Johannesd3 commented:
Tbh I am not convinced of all your changes, especially not the big ones.

That operator overloading would be a better choice is just one point of view. One could also argue that "adding" and "subtracting" range set are rather union and intersection of sets, so the symbols don't really fit. Since the old api worked fine, your suggestion is not objectively better, and it's not even public, I would tend to keeping it as it was.

Another change that I don't understand are the changes to `Range`. Using `start` and an unsigned `length` had the advantage that every possible value would be valid. Also I don't quite understand your arguments against making it fields public.

There are some improvements, but the big changes distract from those. It would be much easier to review if the diff would be smaller. 

The documentations and tests are really welcome, you certainly noticed the lack of it in this repository.

My suggestion: Cut down some of your changes. If you still think there are some bigger changes necessary, let's discuss it here and add those back in separate commits.

ThouCheese commented:
Hi, sorry for the delayed response, but thanks for your straightforward review. I will try to reduce the diff by removing the following: operator overloading and the range with `start` and `end`.

ThouCheese commented:
I have reduced the diff somewhat, but a large part remains due to the privacy of start and end. I can remove that as well, but I believe that there is some sense of value in privacy when working with structs, namely:

1. The classical OOP argument of encapsulation. In range set, we can now not care about whether length, or start or end are a field or a function, but instead we are forced to access these 3 properties in a consistent way.
2. We can easily make modifications to the `Range` struct without breaking any further code in range_set.rs or fetch/*.

If you're not convinced I'll take it out and revert to the original implementation with 2 field accesses and 1 function of course, let me know!

Johannesd3 reviewed (commented):
Yes, I'm more satisfied now.

While Rust isn't really an OOP language, I think it's ok to use methods here.

You need to rebase your changes, it's not mergeable right now, and I want to see whether the checks pass.

roderickvd commented:
@ThouCheese can you rebase so we can let the checks run and merge on success?

ThouCheese commented:
Hi, thanks for the ping! I had totally forgotten about this change. I will take a look at the @Johannesd3 left, resolve them and rebase.

ThouCheese commented on audio/src/range_set.rs:
```diff
@@ -1,124 +1,132 @@
+    /// Checks whether this `RangeSet` contains no values.
     pub fn is_empty(&self) -> bool {
-        self.ranges.is_empty()
+        self.iter().all(Range::is_empty)
```
Is you are right, and the old version would be more performant too.

ThouCheese commented on audio/src/range_set.rs:
```diff
@@ -1,124 +1,132 @@
+mod range;
+pub(crate) use range::Range;
+
+/// An ordered set of non overlapping ranges.
+#[derive(Clone, Default, Debug, PartialEq)]
+pub struct RangeSet {
+    /// For the implementation it is assumed that `ranges` is ordered by `range.start()`.
+    ranges: Vec<Range>,
```
These second two invariants are certainly upheld, but they are not assumed for the code to work. It is a good idea to document them, as they help improve the performance of `RangeSet`, so communicating this to future maintainers is certainly useful.

ThouCheese commented on audio/src/range_set.rs:
```diff
@@ -1,124 +1,132 @@
-                return 0;
-            } else if range.start <= value && value < range.end() {
-                return range.end() - value;
-            }
-        }
-        0
+        self.iter()
+            .take_while(|r| value >= r.start()) // stop once the ranges no longer contain this value
+            .find(|r| r.contains(value)) // find the first range that contains this value
+            .map(|r| r.end() - value) // calculate the remaining elements in that range
+            .unwrap_or(0) // return zero otherwise
     }
```
Do you mean that the old implementation makes more sense or that the new implementation makes more sense, and why?

Johannesd3 replied:
Suppose the passed value is in a range, but there's another range that follows directly after this range. This function would return only consider the first range. That's a "bug" in both implementations, or rather requires a (formerly) undocumented invariant.

ThouCheese replied:
Right! Then I will be more explicit about the requirement in the documentation, and mark this conversation as resolved.

ThouCheese commented:
I have resolved all comments but one, which I did not fully understand, and rebased.

Johannesd3 commented on audio/src/range_set.rs:
```diff
@@ -1,124 +1,138 @@
+
+/// An ordered set of ranges.
+///
+/// ### Invariants:
+/// 1. `ranges` is ordered by `range.start()`. This *must* be  upheld for the implementation to be
+/// correct.
+/// 2. The ranges do not touch. Upon insertion of a range that touches another, instead, those
+/// ranges are merged. This *must* be  upheld for the implementation to be correct.
+/// 3. None of the ranges are empty. Adding empty ranges causes needless performance degradation.
```
Please don't differentiate between different "strengths" of invariants. Just enforce that they are always uphold for code outside of this module.

If these invariants do not affect the use of the public api, they should not be documented as rustdoc but as normal comments inside of the struct.

> None of the ranges are empty.

Strictly speaking, this is a direct consequency of the 2. invariant: An empty range touches itself. However, this might be not directly obvious, so it's good to have it documented.

> Adding empty ranges causes needless performance degradation.

There's no need to worry much about performance. I hardly believe it would affect librespot in a measurable way.  If you still want to consider it, you should set up benchmarks for different use cases. 

(One possibility that _could_ be faster: A `BTreeMap` with "start" as key and "length" as value. Many operations would be O(log n), although there's no way to merge two maps in the most efficient way without accessing its internals. So probably it's not really worthwhile.)

ThouCheese replied:
> Please don't differentiate between different "strengths" of invariants.

Gone.

> An empty range touches itself.

If we define `touching` of two sets S1 and S2 as stating that there exists an element b in the boundary 𝛿S1 of S1 such that b ∈ closure(S2), then since the boundary 𝛿Ø of the empty set Ø is again Ø, no set touches the empty set, including the empty set itself.  Though I admit this is a bit theoretical to be immediately obvious.

> There's no need to worry much about performance.

That is fair, I will remove the comment

Johannesd3 replied:
Well, it's really a matter of definition. First of all, you don't define of what kind of sets you're speaking of. Sets of natural numbers don't make any sense, because on discrete topologies, boundary and closure of a set are the set itself. So you're speaking of half-open sets on real numbers?

But whatever theoretical definition you choose, here's the implemention of `touches` in the code:
```rust
self.start() <= other.end() && self.end() >= other.start()
```

The empty range touches itself.

ThouCheese replied:
You are right. But should we really leave the third invariant implicit?

Johannesd3 replied:
> But should we really leave the third invariant implicit?

No, of course not :)

Johannesd3 commented on audio/src/range_set.rs:
```diff
@@ -1,124 +1,138 @@
-            }
-        }
-        false
+        self.iter().any(|r| r.contains(value))
```
If you want to (it's really not necessary), take a look at https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search_by

ThouCheese replied:
Yes I binary search would be faster since our data is ordered. Good idea! I have implemented this with a closure containing a match, let me know what you think.

Johannesd3 commented on audio/src/fetch/receive.rs:
```diff
@@ -238,24 +238,24 @@ impl AudioFileFetch {
-                let range = tail_end.get_range(0);
-                let offset = range.start;
-                let length = min(range.length, bytes_to_go);
+                let range = tail_end[0];
```
TBH I liked the old version more. I think it's not really obvious what kind of things are indexed, if you understand what I mean.

ThouCheese replied:
You mean the explicitness of `get_range` as opposed to indexing the `RangeSet`?

Johannesd3 commented on audio/src/range_set/range.rs:
```diff
@@ -0,0 +1,305 @@
+    }
+}
+
+impl From<std::ops::Range<usize>> for Range {
+    fn from(r: std::ops::Range<usize>) -> Self {
+        assert!(r.start <= r.end);
+        Self {
+            start: r.start,
+            len: r.end - r.start,
+        }
+    }
```
https://doc.rust-lang.org/std/convert/trait.From.html:

> **Note: This trait must not fail.** If the conversion can fail, use `TryFrom`.

And with `TryFrom` it's getting too verbose ... maybe drop this implementation and use the good old `fn new` wherever possible.

ThouCheese replied:
I will use TryInto, and then add a comment that explains why it will not panic whenever I unwrap it. Does that sound good?

ThouCheese replied:
I see now that I did not read your comment well, I will look into it tomorrow.

ThouCheese replied:
@Johannesd3 Alternatively if a range that is given in descending order is constructed, I can reorder `begin` and `end` to get an infallible conversion.

Johannesd3 replied:
I'm not sure. It's just a matter of fact that this range and std's range are not compatible, so it's probably not good to use the syntactic sugar for ranges to create our Range.

JasonLG1979 commented:
The OCD in me sees this PR open every time I go to the librespot repo and wonders if it will ever be merged? It seems to have stalled out 6 months ago?

roderickvd commented:
I hope @Johannesd3 returns.

roderickvd commented:
I am sorry that this went into disrepair and the work that you put into this.

Cleaning up, I will close for now, but will be absolutely happy to go forward with this if you or others would update it.
